### DISCUSSION TITLE
Events not clearing when Alpine is removed.

### DISCUSSION THREAD

**Discussion by BigWillie** [alpinejs/alpine#3778] [1. Help] [answered]
My first time using Alpine. Seeing if I can integrate Alpine js into an old and massive jQuery and pure DOM code base (no templates, nothing declarative or reactive in it).

My approach is - Alpine template as es6 string

```
const myTemplate = `<div x-data>
    <button @click="$store.test.open = !$store.test.open">Expand</button>
    <span x-show="$store.test.open">
        Content...
    </span>
</div>`
```

I pass this template string into an existing jQuery dialog.

```
  $.dialog({
  ...
    title: dialogTitle,
    content: myTemplate,
  ...  

```
When I load the dialog, I initialise Alpine and the Alpine store. This works correctly. Outside, in jQuery land I can interact with the store via Alpine.store('test') and so on. After closing the dialog, all DOM elements (alpine components in the myTemplate string) are removed.

However, upon reopening the dialog, an issue arises. Clicking the button triggers the event twice, toggling the `$store.test.open` value from true to false. On closing and reopening the dialog multiple times, the number of event triggers increases:

1st time: Once
2nd time: Twice
3rd time: Three times and so on..

This pattern repeats for any additonal buttons that I add to the markup, indicating that the Alpine events aren't being cleared when the DOM elements are removed. The events just stack up. Is there a built-in Alpine method I can use to remove these events when the dialog closes, or, do I need to implement a custom solution?

**Comment by ekwoka** [comment, marked as answer]
I believe you're calling alpine start each time you open the thing. This makes there be multiple "instances" of alpine watching the button. The event listeners are removed when the element is (even if you store the element and put it back) but you ahve multiple being added.

Also, just stop using jquery.

Use modern dom apis :)

**Reply by BigWillie** [reply]
Thanks! So there is no Alpine.stop  method. So I just need to place start somewhere sensible. I'm assuming that calling start binds the events again?

Edit - I just looked up Alpine.start and it says it should fire a warnrng if called more than once? I don't see the warning. The problem does appear to be fixed however when I move start. Thanks!

I would love to use modern DOM apis. The code base which has evolved over 17 years , has grown to hundreds of thousands of lines. Initially developed with pure DOM scripting, it later integrated jQuery, and has since continued to evolve. Indeed, parts of the code base do use modern DOM apis. We have CSS floats from a decade ago and CSS grid...

The product is trusted by thousands of people, and remains a market leader.  I recently joined the company. We are a dev team of 2. We have the responsibility to maintain and enhance this product without blowing it up.

From experience, I know a rewrite is out of the question. It will inevitably end in failure. We want to take a Triggers broom (Ship of Theseus) approach: incremental changes, replace small portions, identify issues and test. We can TDD it out.

Goals:
1. Shift away from direct DOM manipulation. Our focus should be on data manipulation, rather than direct DOM interactions.
2. Future proofing. We want to use that which is established and well documented. The latest hot JS framework is not for us.
3. Compatibility: Any new integration must coexist with our existing code.

I experimented with web components, but didn't like working with the APIs. Lit looks promising, but I don't trust google not to kill it. I had a go at making web components with Alpine - even managed to get slots working, but hit a couple of snags. 

If starting from scratch, I'd probably go with Vue or React simply because they're both established and have large developer communities - but this is not an option. 

Alpine looks like a good option. It's popularity, especially among the server side framework crowd suggests longevity and community support. It's API reminds me of earlier versions of Vue. It sits well with our existing code. We don't have the option of sending html down the wire - but we are sending own an initial page of HTML. If I can figure out how to make client side re-usable components with Alpine, that would be extremely helpful.
